### PR TITLE
[Extensions]Add query for initialized extensions

### PR DESCRIPTION
**Pull request by mloufra**
Signed-off-by: Frank Lou <mloufra@amazon.com>

### Description
Create a feature to have an extension query to OpenSearch to obtain a list of initialized extensions and determine whether a particular extension is initialized.
Equivalent PR on OpenSearch-sdk-java : https://github.com/opensearch-project/opensearch-sdk-java/pull/300 

### Issues Resolved
fixes https://github.com/opensearch-project/opensearch-sdk-java/issues/149

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 1]
```diff
@@ -24,41 +25,53 @@
...
+        this(requestType, null);
+    }
+
+    public ExtensionRequest(ExtensionsManager.RequestType requestType, Optional<String> uniqueId) {
         this.requestType = requestType;
+        this.uniqueId = uniqueId;
     }
 
     public ExtensionRequest(StreamInput in) throws IOException {
-        super(in);
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:34]: I believe Requests require the `super(in)` while responses do not.  This should definitely match the `writeTo` implementation; I don't expect it will pass tests without it matching.

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 2]
```diff
@@ -24,41 +25,53 @@
...
     private static final Logger logger = LogManager.getLogger(ExtensionRequest.class);
-    private ExtensionsManager.RequestType requestType;
+    private final ExtensionsManager.RequestType requestType;
+    private final Optional<String> uniqueId;
 
     public ExtensionRequest(ExtensionsManager.RequestType requestType) {
+        this(requestType, null);
+    }
+
+    public ExtensionRequest(ExtensionsManager.RequestType requestType, Optional<String> uniqueId) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:35]: Don't make the user provide an optional.  Just take a `@Nullable String` parameter here, and then set `this.uniqueId` the same way you do in the stream constructor.  Given the overloaded constructor above, this would set the optional to null and break things later.

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 3]
```diff
@@ -24,41 +25,53 @@
...
+        return "ExtensionRequest{" + "requestType=" + requestType + "uniqueId=" + uniqueId + '}';
     }
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         ExtensionRequest that = (ExtensionRequest) o;
-        return Objects.equals(requestType, that.requestType);
+        return Objects.equals(requestType, that.requestType) && uniqueId.equals(that.uniqueId);
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:70]: (Style) use the same comparison format (either `Objects.equals()` or direct equals comparison) for both.

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 4]
```diff
@@ -24,41 +25,53 @@
  */
 public class ExtensionRequest extends TransportRequest {
     private static final Logger logger = LogManager.getLogger(ExtensionRequest.class);
-    private ExtensionsManager.RequestType requestType;
+    private final ExtensionsManager.RequestType requestType;
+    private final Optional<String> uniqueId;
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:29]: It looks like there is no test class for `ExtensionRequest` so you should create one and include a test of this uniqueID, both in the `null` case (single arg constructor) or value present case (two arg constructor)

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Great progress.  A few more tweaks needed.

File: server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java [hunk 5]
```diff
@@ -528,6 +529,86 @@ public void testHandleExtensionRequest() throws Exception {
         assertEquals("Handler not present for the provided request", exception.getMessage());
     }
 
+    public void testExtensionRequest() throws Exception {
+        ExtensionsManager.RequestType expectedRequestType = ExtensionsManager.RequestType.REQUEST_EXTENSION_DEPENDENCY_INFORMATION;
+
+        // Test ExtensionRequest 2 arg constructor
+        Optional<String> expectedUniqueId = Optional.of("test uniqueid");
+        ExtensionRequest extensionRequest = new ExtensionRequest(expectedRequestType, expectedUniqueId.orElse(null));
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java:537]: This works but I think it would be more readable and conform to the method signature as:
```java
String expectedUniqueId = "test uniqueid";
ExtensionRequest extensionRequest = new ExtensionRequest(expectedRequestType, expectedUniqueId);
// and when tested on lines 539 and 548...
assertEquals(Optional.of(expectedUniqueID), extensionRequest.getUniqueId());
```

File: server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java [hunk 6]
```diff
@@ -528,6 +529,86 @@ public void testHandleExtensionRequest() throws Exception {
...
+            out.flush();
+            try (BytesStreamInput in = new BytesStreamInput(BytesReference.toBytes(out.bytes()))) {
+                extensionRequest = new ExtensionRequest(in);
+                assertEquals(expectedRequestType, extensionRequest.getRequestType());
+                assertEquals(expectedUniqueId, extensionRequest.getUniqueId());
+            }
+        }
+
+        // Test ExtensionRequest 1 arg constructor
+        extensionRequest = new ExtensionRequest(expectedRequestType, null);
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java:553]: Actually use the 1 arg constructor here so you can test the overloaded constructor:
```java
extensionRequest = new ExtensionRequest(expectedRequestType);
```

File: server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java [hunk 7]
```diff
@@ -528,6 +529,86 @@ public void testHandleExtensionRequest() throws Exception {
...
+        assertEquals(expectedRequestType, extensionRequest.getRequestType());
+
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java:555]: Add a test to `assertEquals(Optional.empty(), extensionRequest.getUniqueId());` here and below line 562.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Looks really good!  Some small tweaks left.

Note to other reviewers: previous code review conducted on #5628 in a PR to the (no longer used) feature branch.

**Review by dbwiddis — APPROVED** [review]
LGTM!

**Review by ryanbogan — APPROVED** [review]

File: server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java [hunk 8]
```diff
@@ -82,6 +82,17 @@ public List<ExtensionDependency> getDependencies() {
         return dependencies;
     }
 
+    public boolean dependenciesContain(ExtensionDependency dependency) {
+        for (ExtensionDependency extensiondependency : this.dependencies) {
+            if (dependency.getUniqueId().equals(extensiondependency.getUniqueId())
+                && dependency.getVersion().equals(extensiondependency.getVersion())) {
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java:88]: We don't need if condition here. Logical operator returns a boolean value itself.
```suggestion
                return dependency.getUniqueId().equals(extensiondependency.getUniqueId()) && (dependency.getVersion().equals(extensiondependency.getVersion()) 
```

File: server/src/main/java/org/opensearch/extensions/ExtensionDependencyResponse.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.transport.TransportResponse;
+
+/**
+ * The response for getting the Extension Dependency.
+ *
+ * @opensearch.internal
+ */
+public class ExtensionDependencyResponse extends TransportResponse {
+    private List<DiscoveryExtensionNode> dependencies;
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionDependencyResponse.java:26]:
```suggestion
    private List<DiscoveryExtensionNode> extensionDependencies;
```

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java [hunk 10]
```diff
@@ -79,6 +80,7 @@ public class ExtensionsManager {
     public static final String REQUEST_EXTENSION_ENVIRONMENT_SETTINGS = "internal:discovery/enviornmentsettings";
     public static final String REQUEST_EXTENSION_ADD_SETTINGS_UPDATE_CONSUMER = "internal:discovery/addsettingsupdateconsumer";
     public static final String REQUEST_EXTENSION_UPDATE_SETTINGS = "internal:discovery/updatesettings";
+    public static final String REQUEST_EXTENSION_DEPENDENCY_INFORMATION = "internal:discovery/dependencyinformation";
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:83]: For future: We can change all the action name to `internal:extensions`. @mloufra can you create an issue to address this?

Comment (mloufra) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:83]: sure

File: server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java [hunk 11]
```diff
@@ -82,6 +82,14 @@ public List<ExtensionDependency> getDependencies() {
         return dependencies;
     }
 
+    public boolean dependenciesContain(ExtensionDependency dependency) {
+        for (ExtensionDependency extensiondependency : this.dependencies) {
+            return dependency.getUniqueId().equals(extensiondependency.getUniqueId())
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java:87]: This can return `false` as well, are we fine with it? If not, then if condition was the right way to go about it.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java (hunk 10 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:83]: Can you paste the issue link here?

File: server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java (hunk 11 above)
Comment (mloufra) [reply server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java:87]: Since the return type is `boolean`, We should return either `true` or `false`. Let's keep it the old way.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java (hunk 10 above)
Comment (mloufra) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:83]: https://github.com/opensearch-project/opensearch-sdk-java/issues/305

**Comment by saratvemulapalli** [comment]
@mloufra can you resolve conflicts with changelog.md?

**Review by owaiskazi19 — APPROVED** [review]

File: CHANGELOG.md [hunk 12]
```diff
@@ -21,6 +21,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add support for ppc64le architecture ([#5459](https://github.com/opensearch-project/OpenSearch/pull/5459))
 - Added @gbbafna as an OpenSearch maintainer ([#5668](https://github.com/opensearch-project/OpenSearch/pull/5668))
 - Added support for feature flags in opensearch.yml ([#4959](https://github.com/opensearch-project/OpenSearch/pull/4959))
+- Add query for initialized extensions ([#5658](https://github.com/opensearch-project/OpenSearch/pull/5658))
```
Comment (andrross) [inline CHANGELOG.md:24]: @owaiskazi19 Just a heads up, but since this change was to be backported and released in 2.x, this changelog entry should have gone in the `[Unreleased 2.x]` section of this file. I'm working on cleaning up the changelog on main right now.
